app/assets/javascripts/discourse/app/components/composer-editor.js

```diff
@@ -554,374 +549,402 @@ export default Component.extend(ComposerUploadUppy, {
-          refresh = true;
+        toolbar.addButton({
+          id: "options",
+          group: "extras",
+          icon: "cog",
+          title: "composer.options",
+          sendAction: this.onExpandPopupMenuOptions.bind(this),
+          popupMenu: true,
+        });
+      },
+
+      previewUpdated(preview) {
+        // cache jquery objects for functions still using jquery
+        const $preview = $(preview);
+
+        // Paint mentions
+        const unseenMentions = linkSeenMentions(preview, this.siteSettings);
+        if (unseenMentions.length) {
+          discourseDebounce(
+            this,
+            this._renderUnseenMentions,
+            preview,
+            unseenMentions,
+            450
+          );
         }
 
-        const paintedCount = loadOneboxes(
-          preview,
-          ajax,
-          this.get("composer.topic.id"),
-          this.get("composer.category.id"),
-          this.siteSettings.max_oneboxes_per_post,
-          refresh
-        );
+        this._warnMentionedGroups(preview);
+        this._warnCannotSeeMention(preview);
 
-        if (refresh && paintedCount > 0) {
-          post.set("refreshedPost", true);
+        // Paint category, tag, and other data source hashtags
+        let unseenHashtags;
+        const hashtagContext =
+          this.site.hashtag_configurations["topic-composer"];
+        if (this.siteSettings.enable_experimental_hashtag_autocomplete) {
+          unseenHashtags = linkSeenHashtagsInContext(hashtagContext, preview);
+        } else {
+          unseenHashtags = linkSeenHashtags(preview);
+        }
+        if (unseenHashtags.length > 0) {
+          discourseDebounce(this, this._renderUnseenHashtags, preview, 450);
         }
-      };
 
-      discourseDebounce(this, paintFunc, 450);
+        // Paint oneboxes
+        const paintFunc = () => {
+          const post = this.get("composer.post");
+          let refresh = false;
 
-      // Short upload urls need resolution
-      resolveAllShortUrls(ajax, this.siteSettings, preview);
+          //If we are editing a post, we'll refresh its contents once.
+          if (post && !post.get("refreshedPost")) {
+            refresh = true;
+          }
 
-      preview.addEventListener("click", this._handleImageScaleButtonClick);
-      this._registerImageAltTextButtonClick(preview);
+          const paintedCount = loadOneboxes(
+            preview,
+            ajax,
+            this.get("composer.topic.id"),
+            this.get("composer.category.id"),
+            this.siteSettings.max_oneboxes_per_post,
+            refresh
+          );
+
+          if (refresh && paintedCount > 0) {
+            post.set("refreshedPost", true);
+          }
+        };
+
+        discourseDebounce(this, paintFunc, 450);
+
+        // Short upload urls need resolution
+        resolveAllShortUrls(ajax, this.siteSettings, preview);
+
+        this._generateVideoThumbnail();
```

Comment:
Since this only works when diff preview is enabled, should we explicitly only run thumbnail generation for videos only if that setting is checked?